framework/src/modules/chain/bft/finality_manager.js

```diff
@@ -200,39 +209,85 @@ class FinalityManager {
 			return false;
 		}
 
-		const higherPairVoted = Object.entries(this.preVotes)
+		const highestHeightPreVoted = Object.keys(this.preVotes)
 			.reverse()
-			.find(pair => pair[1] >= this.preVoteThreshold);
+			.find(key => this.preVotes[key] >= this.preVoteThreshold);
 
-		this.prevotedConfirmedHeight = higherPairVoted
-			? parseInt(higherPairVoted[0], 10)
+		this.prevotedConfirmedHeight = highestHeightPreVoted
+			? parseInt(highestHeightPreVoted, 10)
 			: this.prevotedConfirmedHeight;
 
-		const higherPairCommitted = Object.entries(this.preCommits)
+		const highestHeightPreCommitted = Object.keys(this.preCommits)
 			.reverse()
-			.find(pair => pair[1] >= this.preCommitThreshold);
+			.find(key => this.preCommits[key] >= this.preCommitThreshold);
 
-		this.finalizedHeight = higherPairCommitted
-			? parseInt(higherPairCommitted[0], 10)
+		this.finalizedHeight = highestHeightPreCommitted
+			? parseInt(highestHeightPreCommitted, 10)
 			: this.finalizedHeight;
 
 		return true;
 	}
 
+	/**
+	 * Return the valid height to start the pre-commit
+	 * this method will help to identify the gaps in the blocks
+	 * if delegate forged a block on different chain
+	 *
+	 * @param header
+	 * @return {number}
+	 * @private
+	 */
+	_getValidMinHeightToCommit(header) {
+		const searchTillHeight = Math.max(
+			this.minHeight,
+			header.height - this.processingThreshold,
+		);
+
+		let needleHeight = Math.max(
```

Comment:
Just curious why you choose for the variable name `needleHeight`?

Reply:
Does it refer to latest writing height?

Reply:
Its an old term we learned in school while iterating over the complex structure. That you put a needle as a cursor.